### PR TITLE
Feature resize servergroup, refactor operation validation

### PR DESCRIPTION
**Pull request by drmaas**
closes spinnaker/spinnaker#936

### CODE REVIEW

**Comment by drmaas** [comment]
closes spinnaker/spinnaker#936

**Comment by drmaas** [comment]
@spinnaker/reviewers

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy [hunk 1]
```diff
@@ -518,18 +518,32 @@ abstract class OpenstackClientProvider {
   }
 
   /**
+   * Updates a Spinnaker Server Group (Openstack Heat Stack).
+   * @param stackName
+   * @param heatTemplate
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy]: The params here no longer match the signature. Also, can you comment on the differences between template and subtemplate? Including one being a string, one a map.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/servergroup/ResizeOpenstackAtomicOperationConverter.groovy [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import com.netflix.spinnaker.clouddriver.openstack.deploy.converters.OpenstackAtomicOperationConverterHelper
+import com.netflix.spinnaker.clouddriver.openstack.deploy.description.servergroup.ResizeOpenstackAtomicOperationDescription
+import com.netflix.spinnaker.clouddriver.openstack.deploy.ops.servergroup.ResizeOpenstackAtomicOperation
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperation
+import com.netflix.spinnaker.clouddriver.orchestration.AtomicOperations
+import com.netflix.spinnaker.clouddriver.security.AbstractAtomicOperationsCredentialsSupport
+import org.springframework.stereotype.Component
+
+@OpenstackOperation(AtomicOperations.RESIZE_SERVER_GROUP)
+@Component("resizeOpenstackDescription")
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/converters/servergroup/ResizeOpenstackAtomicOperationConverter.groovy]: You don't need the component name

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/servergroup/DeployOpenstackAtomicOperationValidator.groovy [hunk 3]
```diff
@@ -1,5 +1,6 @@
 /*
  * Copyright 2016 Veritas Technologies LLC.
+ * Copyright 2016 Target, Inc.
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/servergroup/DeployOpenstackAtomicOperationValidator.groovy]: Copy paste mistake?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/servergroup/DestroyOpenstackAtomicOperationValidator.groovy [hunk 4]
```diff
@@ -16,27 +16,23 @@
...
 import org.springframework.validation.Errors
 
 @OpenstackOperation(AtomicOperations.DESTROY_SERVER_GROUP)
 @Component
-class DestroyOpenstackAtomicOperationValidator extends DescriptionValidator<OpenstackServerGroupAtomicOperationDescription> {
-  @Autowired
-  AccountCredentialsProvider accountCredentialsProvider
+class DestroyOpenstackAtomicOperationValidator extends AbstractOpenstackDescriptionValidator<OpenstackServerGroupAtomicOperationDescription> {
+
+  String context = "openstackServerGroupAtomicOperationDescription"
```
Comment (varikin) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/servergroup/DestroyOpenstackAtomicOperationValidator.groovy]: Please add `destroy` to the context

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/description/servergroup/ResizeOpenstackAtomicOperationDescription.groovy [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import groovy.transform.Canonical
+
+@Canonical
+class ResizeOpenstackAtomicOperationDescription extends OpenstackAtomicOperationDescription {
+
+  String serverGroupName
+  Capacity capacity = new Capacity()
+
+  static class Capacity {
+    int min
```
Comment (lwander) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/description/servergroup/ResizeOpenstackAtomicOperationDescription.groovy]: Is there no concept of target size? How does autoscaling work when there are no guidelines on how to scale?

File: clouddriver-openstack/src/main/java/org/openstack4j/model/heat/builder/StackUpdateBuilder.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2016 Target, Inc.
```
Comment (varikin) [inline clouddriver-openstack/src/main/java/org/openstack4j/model/heat/builder/StackUpdateBuilder.java]: I question our copyright on this, as it is mostly copy paste to add the stack updates

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/securitygroup/DeleteOpenstackSecurityGroupDescriptionValidator.groovy [hunk 7]
```diff
@@ -32,18 +30,13 @@ import org.springframework.validation.Errors
...
-class DeleteOpenstackSecurityGroupDescriptionValidator extends DescriptionValidator<DeleteOpenstackSecurityGroupDescription> {
-
-  static final String CONTEXT = "deleteOpenstackSecurityGroupAtomicOperationDescription"
-
-  @Autowired
-  AccountCredentialsProvider accountCredentialsProvider
+class DeleteOpenstackSecurityGroupDescriptionValidator extends AbstractOpenstackDescriptionValidator<DeleteOpenstackSecurityGroupDescription> {
 
+  String context = "deleteOpenstackSecurityGroupAtomicOperationDescription" +
+    ""
```
Comment (lwander) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/securitygroup/DeleteOpenstackSecurityGroupDescriptionValidator.groovy]: This seems odd

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/servergroup/AbstractServergroupOpenstackAtomicOperationValidator.groovy [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+   * @param parameters
+   * @return
+   */
+  def validateServerGroup(OpenstackAttributeValidator validator, ServerGroupParameters parameters) {
+    String prefix = "serverGroupParameters"
+    parameters.with {
+      validator.validateNotEmpty(instanceType, "${prefix}.instanceType")
+      validator.validateNotEmpty(image, "${prefix}.image")
+      validator.validateNotNull(maxSize, "${prefix}.maxSize")
+      validator.validatePositive(maxSize, "${prefix}.maxSize")
```
Comment (lwander) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/servergroup/AbstractServergroupOpenstackAtomicOperationValidator.groovy]: Pretty sure `validatePositive` checks for `NotNull`

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/servergroup/DeployOpenstackAtomicOperationValidator.groovy (hunk 3 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/servergroup/DeployOpenstackAtomicOperationValidator.groovy]: No - what is our policy when multiple teams make significant changes to a file? 

@lwander any suggestions?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/description/servergroup/ResizeOpenstackAtomicOperationDescription.groovy (hunk 5 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/description/servergroup/ResizeOpenstackAtomicOperationDescription.groovy]: There is no target size. It starts at the minimum.

File: clouddriver-openstack/src/main/java/org/openstack4j/model/heat/builder/StackUpdateBuilder.java (hunk 6 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/java/org/openstack4j/model/heat/builder/StackUpdateBuilder.java]: @lwander thoughts?

**Comment by lwander** [comment]
LGTM, I'll merge once it's squashed

**Pull request merged**